library/psa_crypto.c

```diff
@@ -4698,21 +4698,107 @@ static psa_status_t psa_generate_derived_key_internal(
     size_t storage_size = bytes;
     psa_status_t status;
 
-    if( ! key_type_is_raw_bytes( slot->attr.type ) )
-        return( PSA_ERROR_INVALID_ARGUMENT );
-    if( bits % 8 != 0 )
-        return( PSA_ERROR_INVALID_ARGUMENT );
-    data = mbedtls_calloc( 1, bytes );
-    if( data == NULL )
-        return( PSA_ERROR_INSUFFICIENT_MEMORY );
+    /*
+    * ECC key types require the generation of a private key which is an integer
+    * in the range [1, N - 1], where N is the boundary of the private key domain:
+    * N is the prime p for Diffie-Hellman, or the order of the
+    * curve’s base point for ECC.
+    *
+    * Let m be the bit size of N, such that 2^m > N >= 2^(m-1).
+    * This function generates the private key using the following process:
+    *
+    * 1. Draw a byte string of length ceiling(m/8) bytes.
+    * 2. If m is not a multiple of 8, set the most significant
+    *    (8 * ceiling(m/8) - m) bits of the first byte in the string to zero.
+    * 3. Convert the string to integer k by decoding it as a big-endian byte string.
+    * 4. If k > N - 2, discard the result and return to step 1.
+    * 5. Output k + 1 as the private key.
+    *
+    * This method allows compliance to NIST standards
```

Comment:
Can you give a precise reference? I'm not sure which standard I should check.

Reply:
Oh, OK, found the references in the PSA spec. Still, since we're quoting that spec extensively here, we might as well give the reference too (just the second bullet point, the one that's about ECC).

Reply:
This is taken from the `PSA Crypto API - psa_key_derivation_output_key (function) for ECC`:
https://armmbed.github.io/mbed-crypto/html/api/ops/kdf.html#psa_key_derivation_output_key